TorchSharp/TorchFiles.cs

```diff
@@ -0,0 +1,129 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+using System.Text;
+
+namespace Torch.IO
+{
+    /// <summary>
+    ///   Abstract base class for all Torch files.
+    /// </summary>
+    public abstract partial class File : IDisposable
+    {
+        internal sealed class HType : SafeHandle
+        {
+            public HType(IntPtr preexistingHandle, bool ownsHandle) : base(IntPtr.Zero, ownsHandle)
+            {
+                SetHandle(preexistingHandle);
+            }
+
+            public override bool IsInvalid => handle == (IntPtr)0;
+            // This is just for marshalling
+            internal HType() : base(IntPtr.Zero, true)
+            {
+            }
+
+            [DllImport ("caffe2")]  
+            extern static void THFile_free (IntPtr handle);
+
+            protected override bool ReleaseHandle()
+            {
+                THFile_free(handle);
+                handle = IntPtr.Zero;
+                return true;
+            }
+        }
+        internal HType handle;
+
+        /// <summary>
+        ///   Finalizer
+        /// </summary>
+        ~File()
+        {
+            Dispose(false);
+        }
+
+        /// <summary>
+        ///   Releases the underlying storage.
+        /// </summary>        
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        ///   Implements the .NET Dispose pattern.
+        /// </summary>
+        protected void Dispose(bool disposing)
+        {
+            if (IsOpen)
+            {
+                Close();
+            }
+
+            if (disposing)
+            {
+                handle.Dispose();
+                handle = null;
+            }
+        }
+
+        [DllImport("caffe2")]
```

Comment:
These do not conform that what we are doing in general where the extern function proceed the function calling it.

Reply:
Fixed, although I'm of the opinion that keeping all the externs in one place, and all the regular methods in one increases readability.

Reply:
I totally agree with you. The fact is that we never agreed on a standard format.